crates/kornia-pnp/src/epnp.rs

```diff
@@ -0,0 +1,633 @@
+//! Efficient Perspective-n-Point (EPnP) solver
+//! Paper: https://www.tugraz.at/fileadmin/user_upload/Institute/ICG/Images/team_lepetit/publications/lepetit_ijcv08.pdf
+//! Reference: https://github.com/opencv/opencv/blob/4.x/modules/calib3d/src/epnp.cpp
+
+use crate::ops::{compute_centroid, gauss_newton};
+use crate::pnp::{NumericTol, PnPError, PnPResult, PnPSolver};
+use glam::{Mat3, Mat3A, Vec3};
+use kornia_lie::so3::SO3;
+use kornia_linalg::rigid::umeyama;
+use kornia_linalg::svd::svd3;
+use nalgebra::{DMatrix, DVector, Vector4};
+
+/// Marker type representing the Efficient PnP algorithm.
+pub struct EPnP;
+
+impl PnPSolver for EPnP {
+    type Param = EPnPParams;
+
+    fn solve(
+        points_world: &[[f32; 3]],
+        points_image: &[[f32; 2]],
+        k: &[[f32; 3]; 3],
+        params: &Self::Param,
+    ) -> Result<PnPResult, PnPError> {
+        solve_epnp(points_world, points_image, k, params)
+    }
+}
+
+/// Parameters controlling the EPnP solver.
+#[derive(Debug, Clone, Default)]
+pub struct EPnPParams {
+    /// Shared numeric tolerances.
+    pub tol: NumericTol,
+}
+
+/// Solve Perspective-n-Point (EPnP).
+///
+/// # Arguments
+/// * `points_world` – 3-D coordinates in the world frame, shape *(N,3)* with `N≥4`.
+/// * `points_image` – Corresponding pixel coordinates, shape *(N,2)*.
+/// * `k` – Camera intrinsics matrix.
+///
+/// # Returns
+/// Tuple *(R, t, rvec)* where
+/// * `R` – 3×3 rotation **world → camera**,
+/// * `t` – 3-vector translation,
+/// * `rvec` – Rodrigues axis-angle representation of `R`.
+pub fn solve_epnp(
+    points_world: &[[f32; 3]],
+    points_image: &[[f32; 2]],
+    k: &[[f32; 3]; 3],
+    params: &EPnPParams,
+) -> Result<PnPResult, PnPError> {
+    let n = points_world.len();
+    if n != points_image.len() {
+        return Err(PnPError::MismatchedArrayLengths {
+            left_name: "world points",
+            left_len: n,
+            right_name: "image points",
+            right_len: points_image.len(),
+        });
+    }
+    if n < 4 {
+        return Err(PnPError::InsufficientCorrespondences {
+            required: 4,
+            actual: n,
+        });
+    }
+
+    let cw = select_control_points(points_world);
+
+    let alphas = compute_barycentric(points_world, &cw, params.tol.eps);
+
+    // Build the 2N×12 design matrix M
+    let m_rows = build_m(&alphas, points_image, k)?;
+
+    let m_flat: Vec<f32> = m_rows.iter().flat_map(|row| row.iter()).cloned().collect();
+    let m_mat = DMatrix::<f32>::from_row_slice(2 * n, 12, &m_flat);
+
+    // Null-space of M (4 right-singular vectors associated with smallest singular values)
+    let svd = m_mat.svd(true, true);
+    let v_t = match svd.v_t {
```

Comment:
`let v_t = svd.v_t.ok_or_else(|| PnPError::SvdFailed("Failed to compute V^T".to_string()))?;`

this would be more rust idiomatic